### ISSUE TITLE
Using `ARG --global EARTHLY_GIT_HASH` results in error when using `FROM --pass-args`

### ISSUE THREAD

**Issue by dustyhorizon** [earthly/earthly#3775]
**What went wrong?**
Specifying `ARG --global EARTHLY_GIT_HASH` or any other builtin args outside of the `base` target (i.e. before the first `FROM` directive) causes `value cannot be specified for built-in build arg EARTHLY_GIT_HASH` when using `--pass-args` options for `FROM`, and `DO` directives (possibly others as well) when referring to external Earthfiles.

These external Earthfiles also have `ARG --global EARTHLY_GIT_HASH` set as well.

**What should have happened?**
Earthly should ignore builtin args from being passed in via `--pass-args` as they are implicitly available anyway in the global context.

**Other Helpful Information**
Earthfiles are running `VERSION 0.8` with Earthly 0.8.3, this was a non-issue when using `VERSION 0.7` with Earthly 0.7.23

**Comment by antoniomdk** [comment]
Same issue here. It seems if a target uses a built-in arg and uses --pass-args to invoke other targets, the issue is raised.

**Comment by dustyhorizon** [comment]
I wrote a simple patch to ignore variable names that matches any built-in args instead of erroring out, I can create the PR but I am not sure if this behavior is acceptable by the Earthly team.

```
diff --git a/variables/parse.go b/variables/parse.go
index 9355bb55..805a5f34 100644
--- a/variables/parse.go
+++ b/variables/parse.go
@@ -1,6 +1,7 @@
 package variables

 import (
+       "fmt"
        "os"
        "strings"

@@ -72,13 +73,14 @@ func parseArg(arg string, pncvf ProcessNonConstantVariableFunc, current *Collect
        }
        if hasValue {
                if reserved.IsBuiltIn(name) {
-                       return "", "", errors.Errorf("value cannot be specified for built-in build arg %s", name)
-               }
-               v, err := parseArgValue(name, value, pncvf)
-               if err != nil {
-                       return "", "", err
+                       fmt.Printf("value cannot be specified for built-in build arg %s and will be ignored\n", name)
+               } else {
+                       v, err := parseArgValue(name, value, pncvf)
+                       if err != nil {
+                               return "", "", err
+                       }
+                       return name, v, nil
                }
-               return name, v, nil
        }
        v, ok := current.Get(name, WithActive())
        if !ok {
```

**Comment by alexcb** [comment]
Thanks for taking a deep dive @dustyhorizon 

I think it would be better to tackle it under https://github.com/earthly/earthly/blob/34320d5111d7488bee0ecbc961b4e9d51d289e06/earthfile2llb/converter.go#L1708 and https://github.com/earthly/earthly/blob/34320d5111d7488bee0ecbc961b4e9d51d289e06/earthfile2llb/converter.go#L1901

However, there's some additional issues though that we need to think about, for example consider:

```
VERSION 0.8
FROM alpine

a:
  RUN env | grep EARTHLY

test:
  ARG EARTHLY_GIT_HASH
  ARG EARTHLY_TARGET
  ARG EARTHLY_NOT_A_RESERVED=hello
  BUILD --pass-args +a
```

when running `+test`, I would expect regular non reserved args to be passed in, e.g. the `EARTHLY_NOT_A_RESERVED` argument. so when we get to the `a` target, we should see:

```
                  +a | EARTHLY_NOT_A_RESERVED=hello
```

But I don't know if we should also see the built ins, e.g.

```
                  +a | EARTHLY_TARGET=+test
                  +a | EARTHLY_GIT_HASH=1d4969b125fb12b010684a1e7fd98a1008086022
```

I'm leaning towards yes, since pass args should pass all the args in it's scope (and it's designed to prevent users from having to always re-declare ARGs)

What I don't fully know is what we should expect the value of `EARTHLY_TARGET` to be -- it was defined under the `test` target, but it was passed to `a` -- should its value change to be `EARTHLY_TARGET=a` even though it wasn't defined under `a`? I _think_ so, but am not entirely sure -- it's an edge case we didn't think about.

**Comment by ingwarsw** [comment]
@alexcb Your last comment sees reasonable..
How hard it will be to release that change?
Cause it started to pop up in many places in our pipelines (cause I upgraded to 0.8 by default)..

**Comment by alexcb** [comment]
> How hard it will be to release that change?

I moved it up in our [todo priority](https://github.com/orgs/earthly/projects/3); however we'll gladly accept a PR if anyone is able to get to it sooner.

**Comment by alexcb** [comment]
potential fix in #3872

**Comment by ingwarsw** [comment]
Today I got into same issue in one of our build pipelines..

Simplest pipeline that fails..
```
VERSION 0.8
ARG --global EARTHLY_GIT_HASH

KBT_PREPARE_IMAGE:
    FUNCTION

kbt-prepare:
    FROM alpine
    DO --pass-args +KBT_PREPARE_IMAGE

kbt-image:
    FROM alpine
    BUILD --pass-args +kbt-prepare
```
